main.go

```diff
@@ -42,41 +46,68 @@ func main() {
 		os.Exit(2)
 	}
 
-	buf := bufio.NewReader(os.Stdin)
-	var regArr []*regexp.Regexp
+	if len(flag.Args()) == 0 {
+		loopCount = 1
+	} else {
+		loopCount = len(flag.Args())
+	}
 
-	if *fixedStringsParameter {
-		for i := range flagTest {
-			flagTest[i] = regexp.QuoteMeta(flagTest[i])
+	for u := 0; u < loopCount; u++ {
+		if *fixedStringsParameter {
+			for i := range flagTest {
+				flagTest[i] = regexp.QuoteMeta(flagTest[i])
+			}
 		}
-	}
 
-	if *wordsParameter {
-		for i := range flagTest {
-			flagTest[i] = "\\b(" + flagTest[i] + ")\\b"
+		if *wordsParameter {
+			for i := range flagTest {
+				flagTest[i] = "\\b(" + flagTest[i] + ")\\b"
+			}
 		}
-	}
 
-	if *linesParameter {
-		for i := range flagTest {
-			flagTest[i] = "(?m:^(" + flagTest[i] + ")$)"
+		if *linesParameter {
+			for i := range flagTest {
+				flagTest[i] = "(?m:^(" + flagTest[i] + ")$)"
+			}
 		}
-	}
 
-	if *ignoreCaseParameter {
-		for i := range flagTest {
-			flagTest[i] = "(?i)" + flagTest[i]
+		if *ignoreCaseParameter {
+			for i := range flagTest {
+				flagTest[i] = "(?i)" + flagTest[i]
+			}
 		}
-	}
 
-	for _, i := range flagTest {
-		re, regErr := regexp.Compile(i)
-		if regErr != nil {
-			fmt.Fprintln(os.Stderr, regErr)
-			os.Exit(2)
+		for _, i := range flagTest {
+			re, regErr := regexp.Compile(i)
+			if regErr != nil {
+				fmt.Fprintln(os.Stderr, regErr)
+				os.Exit(2)
+			}
+			regArr = append(regArr, re)
 		}
-		regArr = append(regArr, re)
+
+		if len(flag.Args()) > 0 {
+			file, fileErr = os.Open(flag.Args()[u])
+			if fileErr != nil {
+				fmt.Fprintln(os.Stderr, fileErr)
+				os.Exit(2)
+			}
+			defer file.Close()
```

Comment:
When is `file.Close()` actually called?

Reply:
At the end of the main function?

Reply:
So during the loop you keep all the files open. What happens if you specify more files than `ulimit -n` allows?

Reply:
I could move my file.Close to the end of my for loop and without defer that should fix the problem

Reply:
👍